tests/real_tests.rs

```diff
@@ -75,114 +83,159 @@ fn get_devices(count: u8) -> Option<Vec<RealTestDev>> {
         .unwrap()
         .as_array()
         .unwrap();
-    if devpaths.len() < count as usize {
-        return None;
+
+    // Parse limits to figure out 1 or 2 quantities of block devices to
+    // use in tests, or fail by returning None.
+    let (low_take, maybe_high_take) = {
+        let avail = devpaths.len();
+        match limits {
+            DeviceLimits::Exactly(num) => {
+                if num > avail {
+                    return None;
+                } else {
+                    (num, None)
+                }
+            }
+            DeviceLimits::AtLeast(min) => {
+                if min > avail {
+                    return None;
+                } else if min == avail {
+                    (min, None)
+                } else {
+                    (min, Some(avail))
+                }
+            }
+            DeviceLimits::Range(min, max) => {
+                assert!(min < max);
+                if min > avail {
+                    return None;
+                } else if min == avail {
+                    (min, None)
+                } else {
+                    (min, Some(cmp::min(avail, max)))
+                }
+            }
+        }
+    };
+
+    let low_paths: Vec<String> = devpaths
+        .iter()
+        .take(low_take)
+        .map(|x| x.as_str().unwrap().to_owned())
+        .collect();
+
+    if maybe_high_take.is_none() {
+        return Some(vec![low_paths]);
     }
-    let devices: Vec<RealTestDev> = devpaths
+    let high_take = maybe_high_take.expect("must not be none; we just checked");
+    let high_paths: Vec<String> = devpaths
         .iter()
-        .take(count as usize)
-        .map(|x| RealTestDev::new(x.as_str().unwrap()))
+        .take(high_take)
+        .map(|x| x.as_str().unwrap().to_owned())
         .collect();
 
-    Some(devices)
+    Some(vec![low_paths, high_paths])
 }
 
-/// Run test on count real devices.
-fn test_with_spec<F>(count: u8, test: F) -> ()
+/// Run test on real devices, using given constraints. Constraints may result
+/// in multiple invocations of the test, with differing numbers of block
+/// devices.
+fn test_with_spec<F>(limits: DeviceLimits, test: F) -> ()
     where F: Fn(&[&Path]) -> ()
 {
     init_logger();
-    let devices = get_devices(count).unwrap();
-    let device_paths: Vec<&Path> = devices.iter().map(|x| x.as_path()).collect();
-    test(&device_paths);
+    let runs = get_devices(limits).unwrap();
+    for run_paths in runs {
+        let devices: Vec<_> = run_paths.iter().map(|x| RealTestDev::new(x)).collect();
```

Comment:
Can't you unify these two lines, w/ map function ```|x| RealTestDev::new(x).as_path()```?

Reply:
no this doesn't work because as_path returns a reference to something that is about to leave scope.